File: SDK-gcc/future_net/componentDepthFirstSearch.cpp
#include "componentDepthFirstSearch.h"
#include "stronglyConnectedComponents.h"

ComponentDepthFirstSearch::ComponentDepthFirstSearch(Graph *graph) {
    this->graph = graph;
    vertexTable = graph->getVertexTable();
    edgeTable = graph->getEdgeTable();
    vertexNum = graph->getVertexNum();
    
    preVisit = new int[vertexNum];
    preVisitSize = vertexNum;
    
    postVisit = new int[vertexNum];
    postVisitSize = vertexNum;
    
}

void ComponentDepthFirstSearch::clear() {
    visited.reset();
    componented.reset();

//    memset(preVisit, 0, vertexNum * sizeof(int));
//    memset(postVisit, 0, vertexNum * sizeof(int));
    visitCnt = 0;
    components = (new StronglyConnectedComponents(graph))->scc();
}

int* ComponentDepthFirstSearch::dfs() {
    clear();
    for (int vid = 0; vid < vertexNum; vid++) {
        if (!visited[vid])
            explore(vid);
    }
    return postVisit;
}

//只遍历startId作为起点的路径
int* ComponentDepthFirstSearch::dfs(int startId) {
    clear();
    explore(startId);
    return postVisit;
}

//if have a path from src to dst
bool ComponentDepthFirstSearch::isReachable(int from, int to) {
    clear();
    explore(from);
    return postVisit[to];
}

void ComponentDepthFirstSearch::explore(int vid) {
    visited.set(vid);
    //set componented to all vertices in same component
    int ci = -1;    //component index
    if (!componented[vid]) {
        std::set<int> comp;
        for (ci = 0; ci < components->size(); ci++) {
            comp = (*components)[ci];
            if (comp.find(vid) != comp.end()) {
                std::set<int>::iterator it;
                for (it = comp.begin(); it != comp.end(); it++) {
                    componented.set(*it);
                }
                break;
            }
        }
        preVisit[vid] = visitCnt++;
    }
    
    int edgeID;
    for (int i = 0; i < vertexTable[vid].outEdge.size(); i++) {
        edgeID = vertexTable[vid].outEdge[i];
        int uid = edgeTable[edgeID].destinationID;
        if (!visited[uid])
            explore(uid);
    }
    
    if (ci != -1) {
        postVisit[vid] = visitCnt++;
        std::set<int>::iterator it;
        for (it = (*components)[ci].begin(); it != (*components)[ci].end(); it++) {
            postVisit[*it] = postVisit[vid];
        }
    }
}

void ComponentDepthFirstSearch::test(Graph *graph) {
    ComponentDepthFirstSearch *cdfs = new ComponentDepthFirstSearch(graph);
    cdfs->dfs(graph->getSrcID());
    for (int i = 0; i < cdfs->vertexNum; i++) {
        cout << cdfs->postVisit[i] << " ";
    }
    cout << endl;
}

File: SDK-java/code/future_net/src/com/routesearch/algorithm/ComponentDepthFirstSearch.java
package com.routesearch.algorithm;

import com.filetool.util.FileUtil;
import com.routesearch.model.Graph;
import com.routesearch.util.PreHandler;

import java.util.List;
import java.util.Set;

/**
 * Created by sunny on 16/3/18.
 */
public class ComponentDepthFirstSearch {

    private Graph graph;
    private List<Set<Integer>> components;
    private boolean[] visited;
    private boolean[] componented;
    private int[] preVisit;
    private int[] postVisit;
    private int visitCnt;

    public ComponentDepthFirstSearch(Graph graph) {
        this.graph = graph;
    }

    private void clear() {
        int n = graph.getVertexNum();
        visited = new boolean[n];
        componented = new boolean[n];
        preVisit = new int[n];
        postVisit = new int[n];
        visitCnt = 0;
        components = new StronglyConnectedComponents(graph).scc(-1);
    }

    public int[] dfs() {
        clear();
        for (int vid = 0; vid < graph.getVertexNum(); vid++) {
            if (!visited[vid])
                explore(vid);
        }
        return postVisit;
    }

    //只遍历startId作为起点的路径
    public int[] dfs(int startId) {
        clear();
        explore(startId);
        return postVisit;
    }

    //if have a path from src to dst
    public boolean isReachable(int from, int to) {
        clear();
        explore(from);
        return postVisit[to] != 0;
    }

    private void explore(int vid) {
        visited[vid] = true;
        //set componented to all vertices in same component
        int ci = -1;    //component index
        if (!componented[vid]) {
            Set<Integer> comp;
            for (ci = 0; ci < components.size(); ci++) {
                comp = components.get(ci);
                if (comp.contains(vid)) {
                    for (int uid : comp) {
                        componented[uid] = true;
                    }
                    break;
                }
            }
            preVisit[vid] = visitCnt++;
        }

        for (int edgeID : graph.getVertexByID(vid).outEdge) {
            int uid = graph.getEdgeByID(edgeID).destinationID;
            if (!visited[uid])
                explore(uid);
        }

        if (ci != -1) {
            postVisit[vid] = visitCnt++;
            for (int uid : components.get(ci)) {
                postVisit[uid] = postVisit[vid];
            }
        }
    }

    //test
    //arg: ./test_case/case1/topo.csv ./test_case/case1/demand.csv
    public static void main(String[] args) {
        String graphFilePath = args[0];
        String conditionFilePath = args[1];

        // 读取输入文件
        String graphContent = FileUtil.read(graphFilePath, null);
        String conditionContent = FileUtil.read(conditionFilePath, null);

        Graph graph = new Graph(graphContent, conditionContent);
//        PreHandler.phr(graph);
        ComponentDepthFirstSearch cdfs = new ComponentDepthFirstSearch(graph);
        cdfs.dfs(graph.getSrcID());
    }
}
